lib/river/request.ex

```diff
@@ -1,11 +1,35 @@
 defmodule River.Request do
+  @type headers :: [{binary, binary}] | %{binary => binary}
+
   defstruct [
     headers: [{"user-agent", "River/#{River.version}"}],
     uri:     nil,
     method:  :get,
     data:    nil
   ]
 
+  @doc """
+    Builds a request struct returning an error if the request would
+    be invalid.
+  """
+  @spec new(URI.t, atom, any, headers) ::
+    {:ok, Request.t} | {:error, :invalid_uri} | {:error, :invalid_method}
```

Comment:
I added typespecs here. I realize the rest of the codebase doesn't have them but I think it would be a good idea to add them as we go. I think it helps a ton with documentation. That said I can remove them if you would rather not go that direction.